src/librustc/ty/layout.rs

```diff
@@ -974,7 +979,8 @@ impl<'a, 'tcx> LayoutCx<'tcx, TyCtxt<'a, 'tcx, 'tcx>> {
             let mut inverse_memory_index: Vec<u32> = (0..fields.len() as u32).collect();
 
             // Anything with repr(C) or repr(packed) doesn't optimize.
-            let mut optimize = (repr.flags & ReprFlags::IS_UNOPTIMISABLE).is_empty();
+            let mut optimize = (repr.flags & ReprFlags::IS_UNOPTIMISABLE).is_empty()
```

Comment:
Now that this is a bit more complex, I think I'll pull this out to a method on ReprOptions, something like `inhibit_field_reordering_opt`.